lib/Parse/ParseDecl.cpp

```diff
@@ -3427,6 +3427,67 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
     }
     break;
   }
+
+  case DeclAttrKind::ABI: {
+    SourceLoc lParenLoc = Tok.getLoc();
+    if (!consumeIfAttributeLParen()) {
+      diagnose(lParenLoc, diag::attr_expected_lparen, AttrName,
+               DeclAttribute::isDeclModifier(DK));
+      return makeParserSuccess();
+    }
+
+    Decl *abiDecl = nullptr;
+    Status |= parseDecl(/*atStart=*/true, /*ifConfigInAttr=*/true, [&](Decl *D){
+      // Look through a TopLevelCodeDecl.
+      // FIXME: Do we need to reparent these decls to match their counterparts?
+      //        Feels like there might be more to do here.
+      if (auto tlcd = dyn_cast<TopLevelCodeDecl>(D)) {
+        ASSERT(tlcd->getBody()->getNumElements() == 1
+                  && "TopLevelCodeDecl with != 1 element?");
+        D = tlcd->getBody()->getElements().front().dyn_cast<Decl *>();
+        if (!D)
+          return;
+      }
+      // TODO: MacroExpansionDecl?
+
+      // When parsing a compound decl, like a PBD, we'll get called for both the
+      // enclosing decl and the ones inside it. We only want the enclosing decl,
+      // which will always come first.
+      if (!abiDecl)
+        abiDecl = D;
+
+      // However, we want to add an implicit inverse @abi attr to all of the
+      // ValueDecls, since they'll need to be able to find their counterparts.
+      // These attrs initially have no `abiDecl`; it'll be filled in by a call
+      // to `connectToInverse()`.
+      if (isa<ValueDecl>(D)) {
+        auto inverseAttr = new (Context) ABIAttr(nullptr, /*isInverse=*/true,
+                                                 /*implicit=*/true);
```

Comment:
We'd like to avoid this kind of implicit operation in Parser as much as possible. Is it possible to move this (and  `connectToInverse(Decl *)`) to a later stage? E.g. `TypeCheckAttr`.
That would also minimize the bridging interface for ASTGen.